sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -854,7 +854,8 @@ public void processElement(ProcessContext c) {
         } else if (numShardsProvider != null) {
           fixedNumShards = numShardsProvider.get();
         } else {
-          fixedNumShards = null;
```

Comment:
Enforcing something like this at runtime is unfortunate. Any way to enforce at graph construction time?